### PR TITLE
Fix Docker image not using GPU

### PR DESCRIPTION
**Pull request by bwest2397**
As it currently stands, the Docker image that gets built is seemingly unable to use the GPU despite the initial "build" stage of the Dockerfile ostensibly being built with CUDA support (i.e. built using the `nvidia/cuda` base image). As reported in https://github.com/jmorganca/ollama/issues/797, it seems that simply setting the second stage of the Dockerfile to use this same `nvidia/cuda` base image resolves the problem.

### CODE REVIEW

**Comment by vinjn** [comment]
`devel` is a big image, you might consider `runtime` tag for final image.

**Comment by mxyng** [comment]
The premise that `ollama/ollama` cannot use NVIDIA GPU is incorrect. While there are a few preconditions on the system level to running Docker with GPU, the image itself has the necessary dependencies to leverage the GPU hardware once it's passed into the container.

The preconditions for Debian 11, as an example are:
1. docker (required to run docker containers)
1. linux-headers (required to build the nvidia kernel module)
1. nvidia-drivers (required to access the nvidia hardware)
1. nvidia-container-toolkit (required to passthrough nvidia hardware through docker)

Changing the base image for the second stage does not negate any of these preconditions.

Once these are available, the last piece is to pass `--gpus all` when starting the ollama container, e.g. `docker run --rm -it --gpus all -p 11434:11434 ollama/ollama`.

```
$ docker pull ollama/ollama
Using default tag: latest
latest: Pulling from ollama/ollama
aece8493d397: Pull complete
3b9196308e0f: Pull complete
e2c8eb914f5c: Pull complete
Digest: sha256:30c3bc1f84846e9bad8f3fd07a4a8bad7747f6c8bfe29ba9418b49002a315411
Status: Downloaded newer image for ollama/ollama:latest
docker.io/ollama/ollama:latest
$ docker run --rm -it --gpus all --name ollama ollama/ollama
2023/11/28 21:15:03 images.go:784: total blobs: 0
2023/11/28 21:15:03 images.go:791: total unused blobs removed: 0
2023/11/28 21:15:03 routes.go:777: Listening on [::]:11434 (version 0.1.12)
```

Without `--gpus all` and all preconditions, the run will output something like this:

```
$ docker run --rm -it --name ollama ollama/ollama
2023/11/28 21:24:16 images.go:784: total blobs: 0
2023/11/28 21:24:16 images.go:791: total unused blobs removed: 0
2023/11/28 21:24:16 routes.go:777: Listening on [::]:11434 (version 0.1.12)
2023/11/28 21:24:16 routes.go:797: warning: gpu support may not be enabled, check that you have installed GPU drivers: nvidia-smi command failed
```

**Comment by bwest2397** [comment]
@vinjn I agree and plan will push a commit when I have tested that.

@mxyng I have a number of other GPU-enabled containers running in Kubernetes, each with the same basic config as what I've tried to use to host Ollama -- that is, GPUs passed to the pod are recognized by the container and everything works without issues. That is to say, my setup is meeting the preconditions for using a GPU through Docker. While those other images/containers have no issues with picking up and using an allocated GPU, I can't get the base `ollama/ollama` image to do so. However, when I build the image with the changes from this PR, it uses the GPU without any issues and everything works as expected.

I don't know why it works on your machine, but my experience, and that of the person who opened that other ticket, seems to contradict what you're saying.

**Comment by mxyng** [comment]
@bwest2397 Kubernetes is slightly different than vanilla Docker, possibly because it's running another container runtime such as containerd or crio. What provider is your cluster hosted on?

If you read through the linked issue, there are comments confirming the Docker image works as I described. Based on your description of the problem and without additional information about your environment, I can only suggest the installed drivers are incompatible with the provided CUDA libraries. This can be resolved by specifying `PATH` and potentially `LD_LIBRARY_PATH` as seen in the kubernetes [example](https://github.com/jmorganca/ollama/blob/main/examples/kubernetes/gpu.yaml).

It turns out nvidia/cuda:runtime sets both `PATH` and `LD_LIBRARY_PATH` to the values set in the kubernetes example which is why it works. If you want to make that change to the dockerfile, I can approve and merge it.

```
$ docker run --rm -it --gpus all --entrypoint env nvidia/cuda:11.8.0-runtime-ubuntu22.04 | grep -e 'PATH' -e 'LD_LIBRARY_PATH'
PATH=/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LD_LIBRARY_PATH=/usr/local/nvidia/lib:/usr/local/nvidia/lib64
```

**Comment by bwest2397** [comment]
@mxyng Running in my Kubernetes cluster (in AWS, to answer your provider question), using the stock `ollama/ollama:0.1.12` and applying the environment variables from your example, when I try to run `mistral` I get the following error:
```
CUDA error 35 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:484: CUDA driver version is insufficient for CUDA runtime version
current device: 0
2023/11/29 23:53:10 llama.go:436: 35 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:484: CUDA driver version is insufficient for CUDA runtime version
current device: 0
```

This error also occurs without manually setting those environment variables.

This error don't occur when using an image built from this PR, i.e. built using `nvidia/cuda:11.8.0-runtime-ubuntu22.04` instead of `ubuntu:22.04`, in which I likewise don't define any environment variables. Switching base images, as in this PR, appears to resolve the problem/[#797](https://github.com/jmorganca/ollama/issues/797).

**Comment by mxyng** [comment]
> Can you link to any specific messages in the thread which they confirm that? I don't see anything supporting what you're saying.

[Here](https://github.com/jmorganca/ollama/issues/797#issuecomment-1764665424) and [here](https://github.com/jmorganca/ollama/issues/797#issuecomment-1764724606) are examples of users successfully running the Docker Hub images with GPU

> CUDA error 35 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:484: CUDA driver version is insufficient for CUDA runtime version

This is certainly unexpected. CUDA error 35 is [`cudaErrorInsufficientDriver`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g3f51e3575c2178246db0a94a430e0038):
```
This indicates that the installed NVIDIA CUDA driver is older than the CUDA runtime library. This is not a supported configuration. Users should install an updated NVIDIA display driver to allow the application to run.
```

Do you happen to know what GPU and NVIDIA driver version are attached to the node? `nvidia-smi` in a pod should give that information as well as the pod's CUDA version. I'm trying to reproduce this in AWS but I'm unable to using these configurations:

```
AMI type
AL2_x86_64_GPU

AMI release version
1.28.3-20231116

Instance types
g4dn.xlarge

NVIDIA/k8s-device-plugin
v0.14.3
```

If you can describe your configurations to reproduce, perhaps we can fix the issue without needing to pull in `nvidia/cuda:runtime`. If there's no reasonable solution we can definitely change the second stage image but keep in mind the image size will be significantly larger

**Comment by bwest2397** [comment]
@mxyng 

> [Here](https://github.com/jmorganca/ollama/issues/797#issuecomment-1764665424) and [here](https://github.com/jmorganca/ollama/issues/797#issuecomment-1764724606) are examples of users successfully running the Docker Hub images with GPU

Apologies, I thought by "as described" I thought you meant manually defining the `PATH` and `LD_LIBRARY_PATH` variables. FWIW though, the two messages you linked to seem to have been related to people deploying via docker-compose having trouble getting it to pass along access to the GPU. The issue in that ticket seems to be either that GPU doesn't work normally with the stock container, _or_ that it doesn't work with older GPU/CUDA driver versions. It seems the original title of the ticket was "Support GPU on linux and docker" but that you changed it to the present "Support GPU on older NVIDIA GPU and CUDA drivers", so I'm not sure which might really be the case.

---

When I'm using the build from this PR, `nvidia-smi` gives me:
```
| NVIDIA-SMI 470.182.03   Driver Version: 470.182.03   CUDA Version: 11.8     |
```

Running `ollama/ollama:v0.1.12`:
```
| NVIDIA-SMI 470.182.03   Driver Version: 470.182.03   CUDA Version: N/A      |
```

According to [NVIDIA's documentation](https://docs.nvidia.com/deploy/cuda-compatibility/#minor-version-compatibility), this driver version should support any CUDA 11.x versions, so the version 11.8 that you're building against should work fine with my GPU drivers (as is my experience using the image I built from this PR)

---

Applying the `NVIDIA_DRIVER_CAPABILITIES` to `ollama/ollama:v0.1.12`, I can't seem to run things successfully. I get
```
cuBLAS error 15 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:7586
```
with a response JSON of
```json
{"error":"llama runner exited, you may not have enough available memory to run this model"}
```
(along with a `200` status code when perhaps `500` might be better -- I'll open an issue/ticket for that if I can't find an existing one)

Here's the relevant log dump:
```
llm_load_tensors: ggml ctx size =    0.11 MiB
llm_load_tensors: using CUDA for GPU acceleration
llm_load_tensors: mem required  =   70.42 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 35/35 layers to GPU
llm_load_tensors: VRAM used: 3847.55 MiB
..................................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init: offloading v cache to GPU
llama_kv_cache_init: offloading k cache to GPU
llama_kv_cache_init: VRAM kv self = 256.00 MiB
llama_new_context_with_model: kv self size  =  256.00 MiB
llama_build_graph: non-view tensors processed: 740/740
llama_new_context_with_model: compute buffer total size = 159.07 MiB
llama_new_context_with_model: VRAM scratch buffer: 156.00 MiB
llama_new_context_with_model: total VRAM used: 4259.56 MiB (model: 3847.55 MiB, context: 412.00 MiB)
{"timestamp":1701444263,"level":"INFO","function":"main","line":2917,"message":"HTTP server listening","hostname":"127.0.0.1","port":49406}
{"timestamp":1701444263,"level":"INFO","function":"log_server_request","line":2478,"message":"request","remote_addr":"127.0.0.1","remote_port":60460,"status":200,"method":"HEAD","path":"/","params":{}}
2023/12/01 15:24:23 llama.go:493: llama runner started in 2.000851 seconds

cuBLAS error 15 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:7586
current device: 0
2023/12/01 15:24:24 llama.go:436: exit status 1
2023/12/01 15:24:24 llama.go:510: llama runner stopped successfully
```

**Comment by bwest2397** [comment]
@mxyng 

TL;DR: Ollama v0.1.12 (`ollama/ollama:v0.1.12`) seems to have a bug that's preventing me from properly testing it, but adding `NVIDIA_DRIVER_CAPABILITIES=compute,utility` to `ollama/ollama:v0.1.11` seems to get it to accept and use GPU without needing to modify the built image in any way. I guess that means... mission accomplished! This PR might be unnecessary, and it seems you might be able to get along with just building that environment variable into the image by default. In the meantime, I can just add that environment variable to my deployment and use the existing stock image.

---

When I'm using the build from this PR, `nvidia-smi` gives me:
```
| NVIDIA-SMI 470.182.03   Driver Version: 470.182.03   CUDA Version: 11.8     |
```

Running `ollama/ollama:v0.1.12`:
```
| NVIDIA-SMI 470.182.03   Driver Version: 470.182.03   CUDA Version: N/A      |
```

According to [NVIDIA's documentation](https://docs.nvidia.com/deploy/cuda-compatibility/#minor-version-compatibility), this driver version should support any CUDA 11.x versions, so the version 11.8 that you're building against should work fine with my GPU drivers (as is my experience using the image I built from this PR)

---

Applying the `NVIDIA_DRIVER_CAPABILITIES` to `ollama/ollama:v0.1.12`, I can't seem to run things successfully. I get
```
cuBLAS error 15 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:7586
```
with a response JSON of
```json
{"error":"llama runner exited, you may not have enough available memory to run this model"}
```
(along with a `200` status code when perhaps `500` might be better -- I'll open an issue/ticket for that if I can't find an existing one)

Here's the relevant log dump:
```
2023/12/01 15:42:49 llama.go:292: 7618 MB VRAM available, loading up to 46 GPU layers
2023/12/01 15:42:49 llama.go:421: starting llama runner
2023/12/01 15:42:49 llama.go:479: waiting for llama runner to start responding
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: Tesla M60, compute capability 5.2

< cut a lot out from `llama_model_loader` and `llm_load_print_meta`, for brevity >

llm_load_tensors: ggml ctx size =    0.11 MiB
llm_load_tensors: using CUDA for GPU acceleration
llm_load_tensors: mem required  =   70.42 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 35/35 layers to GPU
llm_load_tensors: VRAM used: 3847.55 MiB
..................................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init: offloading v cache to GPU
llama_kv_cache_init: offloading k cache to GPU
llama_kv_cache_init: VRAM kv self = 256.00 MiB
llama_new_context_with_model: kv self size  =  256.00 MiB
llama_build_graph: non-view tensors processed: 740/740
llama_new_context_with_model: compute buffer total size = 159.07 MiB
llama_new_context_with_model: VRAM scratch buffer: 156.00 MiB
llama_new_context_with_model: total VRAM used: 4259.56 MiB (model: 3847.55 MiB, context: 412.00 MiB)
{"timestamp":1701444263,"level":"INFO","function":"main","line":2917,"message":"HTTP server listening","hostname":"127.0.0.1","port":49406}
{"timestamp":1701444263,"level":"INFO","function":"log_server_request","line":2478,"message":"request","remote_addr":"127.0.0.1","remote_port":60460,"status":200,"method":"HEAD","path":"/","params":{}}
2023/12/01 15:24:23 llama.go:493: llama runner started in 2.000851 seconds

cuBLAS error 15 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:7586
current device: 0
2023/12/01 15:24:24 llama.go:436: exit status 1
2023/12/01 15:24:24 llama.go:510: llama runner stopped successfully
```

The container I was running that was built against this PR was built against Ollama v0.1.11. When I re-build it against v0.1.12 and try again, it picks up the GPU but then also encounters that same `cuBLAS error 15`. Seems to be an issue with v0.1.12. I can't seem to find any issues with it, but I'm happy to open one if you think it's a more general problem. Judging by the message sent in the response, I figure it may be related to https://github.com/jmorganca/ollama/issues/1303 but I can't verify that. Of note may be that I'm trying this with `mistral`. I haven't tested other models to see if the same issue occurs, but I would try that if I opened a ticket for better coverage.

Running `ollama/ollama:v0.1.11` and applying `NVIDIA_DRIVER_CAPABILITIES`, I can confirm that it seems to work as expected. In this case, `nvidia-smi` also gives the following:
```
| NVIDIA-SMI 470.182.03   Driver Version: 470.182.03   CUDA Version: 11.4     |
```

Not sure what problems may come from initially building against CUDA 11.8 in the Dockerfile but then using CUDA 11.4 in the container, but everything _does_ appear to be working. So... Woohoo!

**Comment by bwest2397** [comment]
You sure that's the same error? Your post seems to say error 35, but this was error 15. It seems v0.1.12 had some regression that's causing that error 15, since it doesn't happen under v0.1.11

**Comment by pithuene** [comment]
I have the same issue with versions more recent than v0.1.11. I am using an NVIDIA GeForce GTX Titan X with 12GB of VRAM and was getting the "llama runner exited, you may not have enough available memory to run this model" trying to run orca-mini. Found the error "cuBLAS error 15 at /go/src/github.com/jmorganca/ollama/llm/llama.cpp/gguf/ggml-cuda.cu:7630" in the logs too.
Running with `docker --gpus=all --runtime=nvidia -e NVIDIA_DRIVER_CAPABILITIES=compute,utility -e NVIDIA_VISIBLE_DEVICES=all -p 11434:11434 --name ollama ollama/ollama:0.1.11` works though.
